### PR TITLE
Feature/mongo get token

### PR DESCRIPTION
**Pull request by daiLlew**
- Added implementation for `mongo.GetToken`

### CODE REVIEW

File: mongo/token_store.go [hunk 1]
```diff
@@ -32,7 +34,19 @@ func (m *Mongo) StoreToken(ctx context.Context, tkn schema.Token, i schema.Ident
 }
 
 func (m *Mongo) GetIdentityByToken(ctx context.Context, token string) (*schema.Identity, *schema.Token, error) {
-	return nil, nil, nil
+	s := m.Session.Copy()
+	defer s.Close()
+
+	t, err := m.getTokenByID(ctx, token)
+	if err != nil {
+		return nil, nil, err
```
Comment (ian-kent) [inline mongo/token_store.go]: in other services we've wrapped errors to add context, not sure if this is necessary here or if it's a pattern we're not using anymore?

File: mongo/token_store.go [hunk 2]
```diff
@@ -84,8 +98,30 @@ func (m *Mongo) storeNewActiveToken(ctx context.Context, tkn schema.Token) error
...
 
+func (m *Mongo) getTokenByID(ctx context.Context, tokenID string) (*schema.Token, error) {
+	s := m.Session.Copy()
+	defer s.Close()
+
+	queryForToken := bson.M{
+		"token_id": tokenID,
+		"deleted":  false,
+		"expiry_date": bson.M{
+			"$gt": time.Now(),
```
Comment (ian-kent) [inline mongo/token_store.go:109]: is there any value in the caller being able to differentiate between "not found because it's never existed" and "not found because it's expired"?

File: token/expiry_helper.go [hunk 3]
```diff
@@ -18,7 +18,7 @@ var (
 	invalidTimeFMT = "invalid time value, must be gte 0 and lt %d defaulting to 0"
 )
 
-// TimeHepler provides helper functions for calculating token expiry and TTL times.
+// ExpiryHelper provides helper functions for calculating token expiry and TTL times.
```
Comment (ian-kent) [inline token/expiry_helper.go:21]: not really part of this PR, but is there a reason for having an `ExpiryHelper` struct rather than just a function? I feel like I'm missing something

File: token/tokens.go [hunk 4]
```diff
@@ -48,73 +54,64 @@ func (t *Tokens) NewToken(ctx context.Context, identity schema.Identity) (token
...
 		return
 	}
 
 	if ttl, err = t.GetTTL(token); err != nil {
+		token = nil
 		return
 	}
 
 	if err = t.Cache.StoreToken(ctx, token.ID, identity, ttl); err != nil {
+		token = nil
```
Comment (ian-kent) [inline token/tokens.go]: does this mean a cache write failure results in failing to generate a token?

File: token/tokentest/get_tokens_test.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+				So(store.GetIdentityByTokenCalls(), ShouldHaveLength, 1)
+				So(store.GetIdentityByTokenCalls()[0].Token, ShouldEqual, testID)
+			})
+		})
+	})
+}
+
+func TestTokens_GetStoreTokenExpired(t *testing.T) {
+	Convey("given store.GetIdentityByToken returns an expired token", t, func() {
```
Comment (ian-kent) [inline token/tokentest/get_tokens_test.go:120]: is this testing a scenario which isn't possible in the actual code (since iirc getting an expired token returns nil, rather than an expired error? I might have missed something though)

File: token/tokentest/new_tokens_test.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+			TimeHelper: timeHelper,
+			MaxTTL:     testTTL,
+		}
+
+		token, ttl, err := tokens.NewToken(context.Background(), *testIdentity)
+
+		Convey("then the correct error is returned", func() {
+			So(err, ShouldEqual, errTest)
+			So(ttl, ShouldEqual, 0)
+			So(token, ShouldBeNil)
```
Comment (ian-kent) [inline token/tokentest/new_tokens_test.go]: related to earlier comment, but should this actually return the token (since its just a cache error not a DB error)

File: mongo/token_store.go (hunk 2 above)
Comment (daiLlew) [reply mongo/token_store.go:109]: Maybe not the caller but we might find that information useful. It's just extra context.

File: token/tokentest/new_tokens_test.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+			Cache:      cache,
+			Store:      store,
+			TimeHelper: timeHelper,
+			MaxTTL:     testTTL,
+		}
+
+		tkn, ttl, err := tokens.NewToken(context.Background(), *testIdentity)
+
+		Convey("then the correct error is returned", func() {
+			So(err, ShouldEqual, token.ErrTokenExpired)
```
Comment (ian-kent) [inline token/tokentest/new_tokens_test.go:266]: similar to earlier comment, is this possible? (DB query doesn't appear to return expired tokens, or differentiate in error response)

File: token/expiry_helper.go (hunk 3 above)
Comment (daiLlew) [reply token/expiry_helper.go:21]: 2 methods. 1 to get the current time, 1 to get the expiry.

I added this abstraction to allow me to control the current time and expiry time in the test. Doing it this way enabled me to test the logic of the `get TTL` method.

File: token/tokens.go (hunk 4 above)
Comment (daiLlew) [reply token/tokens.go]: Good point, I will change it so if it fails, its logged as an error but carries on. If the DB call fails then return an error.

File: token/tokentest/tokens_getTTL_test.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+		ttl, err := tokens.GetTTL(tkn)
+
+		So(err, ShouldBeNil)
+		So(ttl, ShouldEqual, testTTL)
+		So(timeHelper.NowCalls(), ShouldHaveLength, 1)
+	})
+}
+
+func TestTokens_GetTTL_RemainderLessThatLL(t *testing.T) {
```
Comment (ian-kent) [inline token/tokentest/tokens_getTTL_test.go]: typo in test function name

File: token/tokentest/get_tokens_test.go (hunk 5 above)
Comment (daiLlew) [reply token/tokentest/get_tokens_test.go:120]: As you said it shouldn't ever be possible to get an expired token back as the mongo query already does this check.
That said this struct doesn't know anything about the mongo or its implementation so it doesn't make any assumptions and checks its self. Granted it's probably unnecessary but it adds a layer of protection if that query ever changes or we swap databases.

File: token/tokentest/new_tokens_test.go (hunk 6 above)
Comment (daiLlew) [reply token/tokentest/new_tokens_test.go]: Updated.

File: mongo/token_store.go (hunk 2 above)
Comment (ian-kent) [reply mongo/token_store.go:109]: my comment might have been unclear, it doesn't look like that info is available to the caller, I was wondering if it should be

File: token/tokentest/get_tokens_test.go (hunk 5 above)
Comment (ian-kent) [reply token/tokentest/get_tokens_test.go:120]: 👍

File: token/tokentest/new_tokens_test.go (hunk 7 above)
Comment (daiLlew) [reply token/tokentest/new_tokens_test.go:266]: See above comment.

**Review by ian-kent — APPROVED** [review]
